### PR TITLE
Don't delete `has_many`s from param functions

### PR DESCRIPTION
**Pull request by jsteiner**
If you have included a has_many in your factory or passed it into a
param function, then we return those values. We also recursively apply
the same cleanup that we do to the top level struct.

Closes https://github.com/thoughtbot/ex_machina/issues/93
Closes https://github.com/thoughtbot/ex_machina/issues/169

### CODE REVIEW

File: lib/ex_machina/ecto.ex [hunk 1]
```diff
@@ -90,9 +91,56 @@ defmodule ExMachina.Ecto do
...
+  end
+
+  defp strip_assoc(original_assoc, record, association_name) do
+    case original_assoc do
+      %{__meta__: %{__struct__: Ecto.Schema.Metadata, state: :built}} ->
+        assoc = recursively_strip(original_assoc)
+        Map.put(record, association_name, assoc)
+      %{__struct__: Ecto.Association.NotLoaded} ->
+        Map.delete(record, association_name)
+      _list ->
```
Comment (pdawczak) [inline lib/ex_machina/ecto.ex]: Have you thought about adding empty line before every pattern to be matched? I't might make it a little bit easier to read, especially, when there are multiple patterns (like in this case).

File: test/ex_machina/ecto_test.exs [hunk 2]
```diff
@@ -44,18 +45,38 @@ defmodule ExMachina.EctoTest do
...
+      articles: [],
     }
   end
 
+  test "params_for/2 recursively strips has_many values" do
+    article = TestFactory.build(:article)
+
+    user_params = TestFactory.params_for(:user, articles: [article])
+
+    assert user_params[:articles] == [%{
```
Comment (pdawczak) [inline test/ex_machina/ecto_test.exs]: I appreciate that in these assertions you're trying to express this _nested capabilities_ of what's produced by the factories (in the end, this is what's added in this PR), but would you consider describing _the whole structure_? Something like:

```elixir
assert user_params == %{
  name: "John Doe",
  admin: false,
  articles: [
    %{
      title: "My Awesome Article"
    }
  ],
}
```

could serve documentation purpose in longer term?

Just a thought... 💭

File: test/ex_machina/ecto_test.exs [hunk 3]
```diff
@@ -44,18 +45,38 @@ defmodule ExMachina.EctoTest do
...
+    article = TestFactory.build(:article)
+
+    user_params = TestFactory.params_for(:user, best_article: article)
+
+    assert user_params[:best_article] == %{title: article.title}
+  end
+
+  test "params_for/2 works with maps as has_many values" do
+    article = %{title: "Foobar"}
+    user_params = TestFactory.params_for(:user, articles: [article])
```
Comment (pdawczak) [inline test/ex_machina/ecto_test.exs]: In previous tests there is empty line to keep building possible values of `article` before using it in `:user` factory; is it typo here not including it, or in purpose?

File: test/ex_machina/ecto_test.exs [hunk 4]
```diff
@@ -74,20 +95,19 @@ defmodule ExMachina.EctoTest do
     }
   end
 
-  test "params_with_assocs/2 doesn't try to save has_many fields" do
-    assert has_association_in_schema?(ExMachina.User, :articles)
+  test "params_with_assocs/2 keeps has_many associations" do
+    article = TestFactory.build(:article)
+    user_params = TestFactory.params_with_assocs(:user, articles: [article])
```
Comment (pdawczak) [inline test/ex_machina/ecto_test.exs]: [Same here](https://github.com/thoughtbot/ex_machina/pull/174/files?diff=split#r84583517)

File: test/ex_machina/ecto_test.exs (hunk 3 above)
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: Good question! The line afterward was used to separate the setup from the exercise and verification phases of the tests https://robots.thoughtbot.com/four-phase-test

File: test/ex_machina/ecto_test.exs (hunk 2 above)
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: I think that's a good thought. Personally, I like just testing the associations because it makes it clear what I should focus on

File: test/ex_machina/ecto_test.exs [hunk 5]
```diff
@@ -44,18 +45,38 @@ defmodule ExMachina.EctoTest do
...
+
+  test "params_for/2 recursively strips has_one values" do
+    article = TestFactory.build(:article)
+
+    user_params = TestFactory.params_for(:user, best_article: article)
+
+    assert user_params[:best_article] == %{title: article.title}
+  end
+
+  test "params_for/2 works with maps as has_many values" do
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: What do you think about `params_for/2 works with maps that contain has_many associations`?

File: test/ex_machina/ecto_test.exs [hunk 6]
```diff
@@ -44,18 +45,38 @@ defmodule ExMachina.EctoTest do
...
+    article = TestFactory.build(:article)
+
+    user_params = TestFactory.params_for(:user, articles: [article])
+
+    assert user_params[:articles] == [%{
+      title: article.title,
+    }]
+  end
+
+  test "params_for/2 recursively strips has_one values" do
```
Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: `has_one associations`?

**Comment by paulcsmith** [comment]
A few small comments. LGTM. And thanks for your thoughts @pdawczak!

File: test/ex_machina/ecto_test.exs (hunk 3 above)
Comment (pdawczak) [inline test/ex_machina/ecto_test.exs]: Thank you for your response @paulcsmith!

Yes, _four-phase_ is great patter! In my original comment I was trying to refer to it (though now, when I re-read it, it's not clear - sorry for that).

I was wondering, what's the difference between those two cases, when the [`article` is _built_](https://github.com/thoughtbot/ex_machina/blob/cbc525ec863e3cc05b6b79050e3d2b76ac1255ee/test/ex_machina/ecto_test.exs#L66), and when the [`article` map is prepared _manually_](https://github.com/thoughtbot/ex_machina/blob/cbc525ec863e3cc05b6b79050e3d2b76ac1255ee/test/ex_machina/ecto_test.exs#L74); I have the feeling, that _semantically_ they are exactly the same and as such I'd thought of them to stay in the realm of _setup_ phase (thus, I was asking about the empty line indication).

Comment (paulcsmith) [inline test/ex_machina/ecto_test.exs]: Oh you are right. It seems like there should be a newline in between there. I didn't check the code closely enough. Good comment!

File: test/ex_machina/ecto_test.exs (hunk 2 above)
Comment (pdawczak) [inline test/ex_machina/ecto_test.exs]: Yes, I see your point, thanks 👍 !

Just one little thing - the test description mentions _strips has_many values_ (which sounds like _deleting_ them), while assertion here seems to prove opposite. Unless it's just me - would rewording the description to _preserves has_many values/associations_ make it clearer?

Comment (pdawczak) [inline test/ex_machina/ecto_test.exs]: Sorry, I think I've misunderstood the description!

The `has_many` association of `User`, the `:articles` is expected to be built, but the _recursive associations_ (eg. `:editor` of `Article`) should not - is this the purpose of this test? In such a case, would the test description `"builds one level of has_many association"` be more helpful? And, as a proof, maybe test like:

```elixir
article = TestFactory.build(:article, editor: build(:user)) # (1). I'm setting editor here explicitly...

user_params = TestFactory.params_for(:user, articles: [article])

assert user_params[:articles] == [%{
  title: article.title,
  editor: nil # (2). ... and proof it's been striped as the second level association
}]
```

What do you think? Would this reveal the intention of the functionality?

**Comment by KamilLelonek** [comment]
Hey, this PR has over 2 months now. When can we expect it to be merged guys?

**Comment by vraravam** [comment]
This might also resolve the [bug](https://github.com/thoughtbot/ex_machina/issues/190) that I just raised. Please merge asap!

**Comment by jsteiner** [comment]
@KamilLelonek @vraravam all set. Sorry about the delay!
